opm/autodiff/SimulatorBase.hpp

```diff
@@ -0,0 +1,195 @@
+        ///     output_dir ("output")          output directoty
+        ///     output_interval (1)            output every nth step
+        ///     nl_pressure_residual_tolerance (0.0) pressure solver residual tolerance (in Pascal)
+        ///     nl_pressure_change_tolerance (1.0)   pressure solver change tolerance (in Pascal)
+        ///     nl_pressure_maxiter (10)       max nonlinear iterations in pressure
+        ///     nl_maxiter (30)                max nonlinear iterations in transport
+        ///     nl_tolerance (1e-9)            transport solver absolute residual tolerance
+        ///     num_transport_substeps (1)     number of transport steps per pressure step
+        ///     use_segregation_split (false)  solve for gravity segregation (if false,
+        ///                                    segregation is ignored).
+        ///
+        /// \param[in] grid          grid data structure
+        /// \param[in] geo           derived geological properties
+        /// \param[in] props         fluid and rock properties
+        /// \param[in] rock_comp_props     if non-null, rock compressibility properties
+        /// \param[in] linsolver     linear solver
+        /// \param[in] gravity       if non-null, gravity vector
+        /// \param[in] disgas        true for dissolved gas option
+        /// \param[in] vapoil        true for vaporized oil option
+        /// \param[in] eclipse_state
+        /// \param[in] output_writer
+        /// \param[in] threshold_pressures_by_face   if nonempty, threshold pressures that inhibit flow
+        SimulatorBase(const parameter::ParameterGroup& param,
+                      const Grid& grid,
+                      const DerivedGeology& geo,
+                      BlackoilPropsAdInterface& props,
+                      const RockCompressibility* rock_comp_props,
+                      NewtonIterationBlackoilInterface& linsolver,
+                      const double* gravity,
+                      const bool disgas,
+                      const bool vapoil,
+                      std::shared_ptr<EclipseState> eclipse_state,
+                      OutputWriter& output_writer,
+                      const std::vector<double>& threshold_pressures_by_face);
+
+        /// Run the simulation.
+        /// This will run succesive timesteps until timer.done() is true. It will
+        /// modify the reservoir and well states.
+        /// \param[in,out] timer       governs the requested reporting timesteps
+        /// \param[in,out] state       state of reservoir: pressure, fluxes
+        /// \param[in,out] well_state  state of wells: bhp, perforation rates
+        /// \return                    simulation report, with timing data
+        SimulatorReport run(SimulatorTimer& timer,
+                            ReservoirState& state);
+
+    protected:
+        Implementation& asImp_() { return *static_cast<Implementation*>(this); }
```

Comment:
> (also I do not like the private methods ending in underscore, without really understanding why).

the reason is to immediately see which members/attributes are public and which are not when reading a piece of code. since this convention is used in some places in opm, but not in others, I changed the underscore by an 'l', as you requested.